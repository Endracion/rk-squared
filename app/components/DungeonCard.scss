:local {
  li.completed {
  }

  li.mastered {
    text-decoration: line-through;
  }

  .details {
    font-size: small;
    margin-bottom: 0;

    span {
      display: inline-block;
      width: 8em;
    }
  }
}