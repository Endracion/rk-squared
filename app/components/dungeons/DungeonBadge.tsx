import * as React from 'react';

import { Dungeon } from '../../actions/dungeons';

import * as _ from 'lodash';

const styles = require('./DungeonBadge.scss');

/**
 * Shows dungeon status as a Bootstrap badge
 */
export const DungeonBadge = ({dungeons}: {dungeons: Dungeon[]}) => {
  if (!dungeons || !dungeons.length) {
    return null;
  }
  const total = dungeons.length;
  const mastered = _.sumBy(dungeons, d => +d.isMaster);
  const hasUnlocked = _.find(dungeons, d => d.isUnlocked && !d.isMaster);
  // TODO: Highlight in red if about to expire?
  // TODO: Add Cid missions
  // TODO: Add a tooltip explaining mastered / completed / total syntax
  // TODO: Mark dungeons that are not yet entered but that should be unlocked? (see access-controls branch)
  // TODO: Dungeons that are mastered but have grade rewards should show as incomplete, here and in list
  const classes = `badge ${styles.component} ` + (hasUnlocked ? 'badge-primary' : 'badge-secondary');
  if (mastered === total) {
    return (
      <span className={classes}>
        {mastered} / {mastered} / {total}
      </span>
    );
  } else {
    const completed = _.sumBy(dungeons, d => +d.isComplete);
    const stamina = _.sumBy(dungeons, d => !d.isMaster ? d.totalStamina : 0);

    return (
      <span className={classes}>
        {mastered} / {completed} / {total}
        <br/>
        {stamina} stamina
      </span>
    );
  }
};
