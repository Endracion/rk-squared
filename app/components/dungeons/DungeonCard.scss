:local {
  /* Size of images like
   * http://ffrk.denagames.com/dff/static/lang/ww/compile/en/image/event_intro/common/pict/icon_series_10.png
   */
  $series-icon-width: 460px;
  $series-icon-height: 112px;
  /* We want the icons to come out to about 40px, to match the size of the
   * game-icons we use for world categories.  Taking into account the internal
   * padding of FFRK's series icons, this works out to 42px.
   */
  $scaled-series-icon-size: 42px;

  $icon-size: 40px;

  /* X offset to position the icon against the left edge. */
  $series-icon-left: -($series-icon-width / 2 * $scaled-series-icon-size / $series-icon-height) + $scaled-series-icon-size;

  li.completed {
  }

  li.mastered {
    text-decoration: line-through;
  }

  .details {
    font-size: small;
    margin-bottom: 0;

    span {
      display: inline-block;
      width: 8em;
    }
  }

  .title:not(.seriesIcon) {
    background: no-repeat 23px;
    background-size: auto $icon-size;
  }

  .title.seriesIcon {
    background: no-repeat $series-icon-left;
    background-size: auto $scaled-series-icon-size;

  }

  .titleText {
    padding-left: calc(#{$icon-size} + 1rem);
  }
}