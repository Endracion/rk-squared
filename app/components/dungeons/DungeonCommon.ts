import {
  Dungeon,
  getAllPrizes,
  getAvailablePrizes,
  hasAvailablePrizes,
  PrizeItem,
} from '../../actions/dungeons';

export const rewards = (isAnonymous?: boolean) =>
  (isAnonymous ? 'available' : 'unclaimed') + ' rewards';
export const rewardsTitle = (isAnonymous?: boolean) =>
  (isAnonymous ? 'Available' : 'Unclaimed') + ' Rewards';

interface DungeonInfo {
  completed: boolean;
  mastered: boolean;
  hasPrizes: boolean;
  getPrizes(dungeonOrDungeons: Dungeon | Dungeon[]): PrizeItem[];
}

interface DungeonProcessor {
  getInfo(dungeon: Dungeon): DungeonInfo;
}

export const anonymousProcessor: DungeonProcessor = {
  getInfo(dungeon: Dungeon) {
    return {
      hasPrizes: true,
      completed: false,
      mastered: false,
      getPrizes: getAllPrizes,
    };
  },
};

export const standardProcessor: DungeonProcessor = {
  getInfo(dungeon: Dungeon) {
    const hasPrizes = hasAvailablePrizes(dungeon);
    return {
      hasPrizes,
      completed: dungeon.isComplete,
      mastered: dungeon.isMaster && !hasPrizes,
      getPrizes: getAvailablePrizes,
    };
  },
};

export const getProcessor = (isAnonymous?: boolean) =>
  isAnonymous ? anonymousProcessor : standardProcessor;
