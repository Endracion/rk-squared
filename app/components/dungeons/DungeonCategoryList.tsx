import * as React from 'react';

import * as _ from 'lodash';

import { descriptions, getSorter, World, WorldCategory } from '../../actions/worlds';
import { CollapsibleCard } from '../common/CollapsibleCard';
import { MaybeWrap } from '../common/MaybeWrap';
import DungeonCard from './DungeonCard';
import WorldBadge from './WorldBadge';
import WorldPrizeList from './WorldPrizeList';

const categoryImages: { [category: string]: string } = {
  [WorldCategory.CrystalTower]: require('../../images/game-icons/white-tower.svg'),
  [WorldCategory.Event]: require('../../images/game-icons/book-cover.svg'),
  [WorldCategory.JumpStart]: require('../../images/game-icons/lob-arrow.svg'),
  [WorldCategory.PowerUpMote]: require('../../images/game-icons/orb-direction.svg'),
  [WorldCategory.Magicite]: require('../../images/game-icons/triple-yin.svg'),
  [WorldCategory.Newcomer]: require('../../images/game-icons/big-egg.svg'),
  [WorldCategory.Nightmare]: require('../../images/game-icons/spectre.svg'),
  [WorldCategory.Raid]: require('../../images/game-icons/swords-emblem.svg'),
  [WorldCategory.Realm]: require('../../images/game-icons/closed-doors.svg'),
  [WorldCategory.Record]: require('../../images/game-icons/galleon.svg'),
  [WorldCategory.Renewal]: require('../../images/game-icons/calendar.svg'),
  [WorldCategory.SpecialEvent]: require('../../images/game-icons/star-formation.svg'),
  [WorldCategory.Torment]: require('../../images/game-icons/daemon-skull.svg'),
};

interface Props {
  worlds: {
    [id: number]: World;
  };
  category: WorldCategory;
}

// TODO: Include banner images?
// E.g.:
// http://ffrk.denagames.com/dff/static/lang/ww/compile/en/image/crystal_tower/event_list/banner/7.png
// (determined by crystal_tower_bundles.0.bundle_id???)
// http://ffrk.denagames.com/dff/static/lang/ww/compile/en/image/event/3026.png

/**
 * Worlds grouped by subcategory.
 *
 * Subcategory may be ''.
 */
type WorldsBySubcategory = Array<[string, World[]]>;

/**
 * Gets the list of worlds, sorted and grouped by subcategory, for the given
 * WorldCategory.
 */
function getSortedWorlds(
  worlds: { [id: number]: World },
  category: WorldCategory,
): WorldsBySubcategory | null {
  // Known subcategories and their sort orders
  const subcategories: { [s: string]: number } = {};
  // Worlds sorted by subcategory
  const bySubcategory: { [s: string]: World[] } = {};

  _.forEach(worlds, w => {
    if (w.category === category) {
      // FIXME: Filter out worlds that aren't open?
      const subcategory = w.subcategory || '';
      subcategories[subcategory] = w.subcategorySortOrder || 0;
      bySubcategory[subcategory] = bySubcategory[subcategory] || [];
      bySubcategory[subcategory].push(w);
    }
  });

  if (_.isEmpty(bySubcategory)) {
    return null;
  }
  let result = _.toPairs(bySubcategory);

  // Sort subcategories - by subcategory sort order if possible, and by
  // subcategory name if not.  Blank goes first.
  result = (_.sortBy(result, [
    (i: any) => (i[0] === '' ? -Infinity : subcategories[i[0]]),
    (i: any) => i[0],
  ]) as any) as WorldsBySubcategory;

  // Sort within each subcategory.
  const sorter = getSorter(category);
  for (const i of result) {
    i[1] = sorter(i[1]);
  }

  return result;
}

const DungeonCategoryTitle = ({
  category,
  title,
  worlds,
}: {
  category?: WorldCategory;
  title: string;
  worlds: World[];
}) => (
  <div className="d-flex justify-content-between align-items-center">
    <div>
      {category != null && (
        <img
          src={categoryImages[category]}
          width={40}
          height={40}
          style={{ paddingRight: '0.5em' }}
          alt=""
        />
      )}
      {title}
    </div>
    <WorldBadge worlds={worlds} />
  </div>
);

export class DungeonCategoryList extends React.PureComponent<Props> {
  render() {
    const { worlds, category } = this.props;
    const bySubcategory = getSortedWorlds(worlds, category);
    if (bySubcategory == null) {
      return null;
    }
    const categoryWorlds = _.flatten(bySubcategory.map(([subcategory, subWorlds]) => subWorlds));
    const id = `dungeon-category-${category}`;
    return (
      <CollapsibleCard
        id={id}
        title={() => (
          <DungeonCategoryTitle
            category={category}
            title={descriptions[category]}
            worlds={categoryWorlds}
          />
        )}
      >
        <div className="accordion">
          <WorldPrizeList worlds={categoryWorlds} />
          {bySubcategory.map(([subcategory, subWorlds], i) => (
            <MaybeWrap
              component={CollapsibleCard}
              test={subcategory !== ''}
              id={`${id}-${i}`}
              key={i}
              title={() => <DungeonCategoryTitle title={subcategory} worlds={subWorlds} />}
            >
              <div>
                {subcategory !== '' && <WorldPrizeList worlds={subWorlds} />}
                <div className="accordion">
                  {subWorlds.map((w, j) => (
                    <DungeonCard world={w} key={j} />
                  ))}
                </div>
              </div>
            </MaybeWrap>
          ))}
        </div>
      </CollapsibleCard>
    );
  }
}
