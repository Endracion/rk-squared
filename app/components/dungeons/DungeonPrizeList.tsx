import * as React from 'react';
import { connect } from 'react-redux';

import { Dungeon, getAllPrizes, getAvailablePrizes } from '../../actions/dungeons';
import { ItemType } from '../../data/items';
import { IState } from '../../reducers';
import { PrizeList } from './PrizeList';

import * as _ from 'lodash';

const alwaysShow = (type: ItemType) => type === ItemType.DropItem;

interface Props {
  dungeons: Dungeon[];
  showItemType: { [t in ItemType]: boolean };
  isAnonymous?: boolean;

  [s: string]: any;
}

export class DungeonPrizeList extends React.Component<Props> {
  render() {
    const { dungeons, showItemType, isAnonymous, ...props } = this.props;
    const prizes = isAnonymous ? getAllPrizes(dungeons) : getAvailablePrizes(dungeons);
    const filteredPrizes = _.filter(prizes, p => alwaysShow(p.type) || showItemType[p.type]);
    return <PrizeList {...props} prizes={filteredPrizes} showTooltips={true} />;
  }
}

interface StateProps {
  showItemType: { [t in ItemType]: boolean };
}

interface OwnProps {
  dungeons: Dungeon[];

  [s: string]: any;
}

export default connect<StateProps, {}, OwnProps>((state: IState) => ({
  showItemType: state.prefs.showItemType,
}))(DungeonPrizeList);
