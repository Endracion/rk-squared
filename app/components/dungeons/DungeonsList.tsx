import * as React from 'react';

import classNames from 'classnames';

import { sortOrder, World, WorldCategory } from '../../actions/worlds';
import { DungeonCategoryList } from './DungeonCategoryList';

interface Props {
  className?: string;
  worlds: {
    [id: number]: World;
  };
  isAnonymous?: boolean;
}

export class DungeonsList extends React.PureComponent<Props> {
  render() {
    const { className, worlds, isAnonymous } = this.props;
    return (
      <div className={classNames(className, 'accordion')}>
        {sortOrder.map((category: WorldCategory, i: number) => (
          <DungeonCategoryList
            worlds={worlds}
            category={category}
            isAnonymous={isAnonymous}
            key={i}
          />
        ))}
      </div>
    );
  }
}
