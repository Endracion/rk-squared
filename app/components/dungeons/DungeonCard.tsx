import * as React from 'react';
import { connect } from 'react-redux';
import * as ReactTooltip from 'react-tooltip';

import * as classNames from 'classnames';

import {
  Dungeon,
  formatDifficulty,
  getAvailablePrizes,
  hasAvailablePrizes,
} from '../../actions/dungeons';
import { World } from '../../actions/worlds';
import { LangType } from '../../api/apiUrls';
import { seriesIcon } from '../../data/urls';
import { IState } from '../../reducers';
import { getDungeonsForWorld } from '../../reducers/dungeons';
import { CollapsibleCard } from '../common/CollapsibleCard';
import { DungeonBadge } from './DungeonBadge';
import DungeonPrizeList from './DungeonPrizeList';
import { PrizeList } from './PrizeList';

const styles = require('./DungeonCard.scss');

interface Props {
  world: World;
}

interface ConnectedProps extends Props {
  dungeons: Dungeon[];
}

const DungeonCardTitle = ({ world, dungeons }: { world: World; dungeons: Dungeon[] }) => (
  <span className={styles.titleText}>
    {world.name}
    <DungeonBadge dungeons={dungeons} />
  </span>
);

const DungeonDetails = ({ dungeon }: { dungeon: Dungeon }) => (
  <p className={styles.details}>
    <span>Difficulty {formatDifficulty(dungeon)}</span>
    <span>Stamina {dungeon.totalStamina}</span>
  </p>
);

export const DungeonListItem = ({ dungeon }: { dungeon: Dungeon }) => {
  const classes = classNames({
    [styles.completed]: dungeon.isComplete,
    [styles.mastered]: dungeon.isMaster,
  });
  const id = `dungeon-item-${dungeon.id}`;
  const showTooltip = hasAvailablePrizes(dungeon);
  return (
    <li className={classes}>
      <div data-tip={showTooltip} data-for={id}>
        {dungeon.name}
        {dungeon.isMaster || <DungeonDetails dungeon={dungeon} />}
      </div>
      {showTooltip && (
        <ReactTooltip place="bottom" id={id}>
          <PrizeList prizes={getAvailablePrizes(dungeon)} />
        </ReactTooltip>
      )}
    </li>
  );
};

/**
 * Lists all of the dungeons for a single world.
 */
export class DungeonCard extends React.PureComponent<ConnectedProps> {
  render() {
    const { world, dungeons } = this.props;
    const noMessage = !world.isUnlocked
      ? 'You have not yet entered these dungeons.'
      : !dungeons
      ? 'These dungeons have not been loaded.'
      : undefined;

    // For speed and simplicity, hard-code icons for GL.
    const icon = world.iconUrl || seriesIcon(LangType.Gl, world.seriesId);

    // TODO: Rework tooltips?  They may be faster if we only render 1 per card.
    // See TormentGrid for the newer approach I'm using.
    return (
      <CollapsibleCard
        id={`world-${world.id}-dungeons`}
        title={() => <DungeonCardTitle world={world} dungeons={dungeons} />}
        titleClassName={classNames(styles.title, { [styles.seriesIcon]: !world.iconUrl })}
        titleStyle={{ backgroundImage: icon ? `url(${icon}` : undefined }}
      >
        {noMessage ? (
          <p className="mb-0">{noMessage}</p>
        ) : (
          <div className="row">
            <div className="col-lg">
              <h6>Dungeons</h6>
              <ul className="mb-0">
                {dungeons.map((d, i) => (
                  <DungeonListItem dungeon={d} key={i} />
                ))}
              </ul>
            </div>
            <div className="col-lg">
              <h6>Unclaimed Rewards</h6>
              <DungeonPrizeList dungeons={dungeons} className="mb-0" />
            </div>
          </div>
        )}
      </CollapsibleCard>
    );
  }
}

export default connect((state: IState, ownProps: Props) => ({
  dungeons: getDungeonsForWorld(state.dungeons, ownProps.world.id),
}))(DungeonCard);
