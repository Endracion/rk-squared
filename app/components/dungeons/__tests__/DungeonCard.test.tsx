import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { DungeonListItem } from '../DungeonCard';

import { Dungeon } from '../../../actions/dungeons';
import { ItemType } from '../../../data/items';

const dungeon: Dungeon = {
  closedAt: 2145938400,
  difficulty: 240,
  id: 15048601,
  isComplete: false,
  isMaster: false,
  isUnlocked: true,
  name: 'Tyranny of the Impure, Part 1',
  openedAt: 1531357200,
  prizes: {
    claimedGrade: [],
    completion: [
      {
        amount: 1500,
        id: 92000000,
        name: 'Gil',
        type: ItemType.Common,
      },
    ],
    firstTime: [
      {
        amount: 1,
        id: 95001020,
        name: 'Memory Crystal III Lode',
        type: ItemType.Common,
      },
      {
        amount: 100000,
        id: 92000000,
        name: 'Gil',
        type: ItemType.Common,
      },
      {
        amount: 1,
        id: 91000000,
        name: 'Mythril',
        type: ItemType.Common,
      },
    ],
    mastery: [
      {
        amount: 1,
        id: 95001020,
        name: 'Memory Crystal III Lode',
        type: ItemType.Common,
      },
      {
        amount: 2,
        id: 40000066,
        name: 'Power Crystal',
        type: ItemType.Orb,
      },
      {
        amount: 100000,
        id: 92000000,
        name: 'Gil',
        type: ItemType.Common,
      },
    ],
    unclaimedGrade: [
      {
        amount: 20,
        id: 95001080,
        name: 'Record Rubies',
        type: ItemType.Common,
      },
      {
        amount: 100000,
        id: 92000000,
        name: 'Gil',
        type: ItemType.Common,
      },
      {
        amount: 30,
        id: 130300300,
        name: 'Vitality Mote (5★)',
        type: ItemType.Mote,
      },
      {
        amount: 100000,
        id: 92000000,
        name: 'Gil',
        type: ItemType.Common,
      },
      {
        amount: 30,
        id: 130300300,
        name: 'Vitality Mote (5★)',
        type: ItemType.Mote,
      },
      {
        amount: 20,
        id: 95001080,
        name: 'Record Rubies',
        type: ItemType.Common,
      },
      {
        amount: 7,
        id: 40000060,
        name: 'Major Holy Orb',
        type: ItemType.Orb,
      },
      {
        amount: 30,
        id: 130300300,
        name: 'Vitality Mote (5★)',
        type: ItemType.Mote,
      },
      {
        amount: 2,
        id: 40000060,
        name: 'Major Holy Orb',
        type: ItemType.Orb,
      },
      {
        amount: 30,
        id: 130400200,
        name: 'Wisdom Mote (4★)',
        type: ItemType.Mote,
      },
      {
        amount: 5,
        id: 130300300,
        name: 'Vitality Mote (5★)',
        type: ItemType.Mote,
      },
    ],
  },
  seriesId: 150001,
  staminaList: [0],
  totalStamina: 1,
};

describe('DungeonCard', () => {
  describe('DungeonListItem', () => {
    it('renders complex prize lists', () => {
      const tree = renderer.create(<DungeonListItem dungeon={dungeon} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('renders complex prize lists in anonymous view', () => {
      const tree = renderer
        .create(<DungeonListItem dungeon={dungeon} isAnonymous={true} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
