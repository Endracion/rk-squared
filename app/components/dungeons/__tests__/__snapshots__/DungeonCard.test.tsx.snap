// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DungeonCard DungeonListItem renders complex prize lists 1`] = `
<li
  className=""
>
  <div
    data-for="dungeon-item-15048601"
    data-tip={true}
  >
    Tyranny of the Impure, Part 1
    <p
      className="details"
    >
      <span>
        Difficulty 
        240
      </span>
      <span>
        Stamina 
        1
      </span>
    </p>
  </div>
  <div
    className="__react_component_tooltip place-top type-dark "
    data-id="tooltip"
    id="dungeon-item-15048601"
  >
    <ul
      className="component"
    >
      <li
        data-for={undefined}
        data-tip={undefined}
      >
        <img
          height={32}
          src="http://ffrk.denagames.com/dff/static/lang/ww/compile/en/image/common_item/95001020.png"
          width={32}
        />
        Memory Crystal III Lode
         ×
        1
      </li>
      <li
        data-for={undefined}
        data-tip={undefined}
      >
        <img
          height={32}
          src="http://ffrk.denagames.com/dff/static/lang/ww/compile/en/image/common_item/92000000.png"
          width={32}
        />
        Gil
         ×
        100000
      </li>
      <li
        data-for={undefined}
        data-tip={undefined}
      >
        <img
          height={32}
          src="http://ffrk.denagames.com/dff/static/lang/ww/compile/en/image/common_item/91000000.png"
          width={32}
        />
        Mythril
         ×
        1
      </li>
    </ul>
    <ul
      className="component"
    >
      <li
        data-for={undefined}
        data-tip={undefined}
      >
        <img
          height={32}
          src="http://ffrk.denagames.com/dff/static/lang/ww/compile/en/image/common_item/95001020.png"
          width={32}
        />
        Memory Crystal III Lode
         ×
        1
      </li>
      <li
        data-for={undefined}
        data-tip={undefined}
      >
        <img
          height={32}
          src="http://ffrk.denagames.com/dff/static/lang/ww/compile/en/image/ability_material/40000066/40000066_112.png"
          width={32}
        />
        Power Crystal
         ×
        2
      </li>
      <li
        data-for={undefined}
        data-tip={undefined}
      >
        <img
          height={32}
          src="http://ffrk.denagames.com/dff/static/lang/ww/compile/en/image/common_item/92000000.png"
          width={32}
        />
        Gil
         ×
        100000
      </li>
    </ul>
    <ul
      className="component"
    >
      <li
        data-for={undefined}
        data-tip={undefined}
      >
        <img
          height={32}
          src="http://ffrk.denagames.com/dff/static/lang/ww/compile/en/image/common_item/95001080.png"
          width={32}
        />
        Record Rubies
         ×
        20
      </li>
      <li
        data-for={undefined}
        data-tip={undefined}
      >
        <img
          height={32}
          src="http://ffrk.denagames.com/dff/static/lang/ww/compile/en/image/common_item/92000000.png"
          width={32}
        />
        Gil
         ×
        100000
      </li>
      <li
        data-for={undefined}
        data-tip={undefined}
      >
        <img
          height={32}
          src="http://ffrk.denagames.com/dff/static/lang/ww/compile/en/image/sphere_material/130300300/130300300_112.png"
          width={32}
        />
        Vitality Mote (5★)
         ×
        30
      </li>
      <li
        data-for={undefined}
        data-tip={undefined}
      >
        <img
          height={32}
          src="http://ffrk.denagames.com/dff/static/lang/ww/compile/en/image/common_item/92000000.png"
          width={32}
        />
        Gil
         ×
        100000
      </li>
      <li
        data-for={undefined}
        data-tip={undefined}
      >
        <img
          height={32}
          src="http://ffrk.denagames.com/dff/static/lang/ww/compile/en/image/sphere_material/130300300/130300300_112.png"
          width={32}
        />
        Vitality Mote (5★)
         ×
        30
      </li>
      <li
        data-for={undefined}
        data-tip={undefined}
      >
        <img
          height={32}
          src="http://ffrk.denagames.com/dff/static/lang/ww/compile/en/image/common_item/95001080.png"
          width={32}
        />
        Record Rubies
         ×
        20
      </li>
      <li
        data-for={undefined}
        data-tip={undefined}
      >
        <img
          height={32}
          src="http://ffrk.denagames.com/dff/static/lang/ww/compile/en/image/ability_material/40000060/40000060_112.png"
          width={32}
        />
        Major Holy Orb
         ×
        7
      </li>
      <li
        data-for={undefined}
        data-tip={undefined}
      >
        <img
          height={32}
          src="http://ffrk.denagames.com/dff/static/lang/ww/compile/en/image/sphere_material/130300300/130300300_112.png"
          width={32}
        />
        Vitality Mote (5★)
         ×
        30
      </li>
      <li
        data-for={undefined}
        data-tip={undefined}
      >
        <img
          height={32}
          src="http://ffrk.denagames.com/dff/static/lang/ww/compile/en/image/ability_material/40000060/40000060_112.png"
          width={32}
        />
        Major Holy Orb
         ×
        2
      </li>
      <li
        data-for={undefined}
        data-tip={undefined}
      >
        <img
          height={32}
          src="http://ffrk.denagames.com/dff/static/lang/ww/compile/en/image/sphere_material/130400200/130400200_112.png"
          width={32}
        />
        Wisdom Mote (4★)
         ×
        30
      </li>
      <li
        data-for={undefined}
        data-tip={undefined}
      >
        <img
          height={32}
          src="http://ffrk.denagames.com/dff/static/lang/ww/compile/en/image/sphere_material/130300300/130300300_112.png"
          width={32}
        />
        Vitality Mote (5★)
         ×
        5
      </li>
    </ul>
  </div>
</li>
`;
